### PR TITLE
SALTO-2895: support zendesk custom statuses

### PR DESCRIPTION
**Pull request by edenhassid**
_support deploy of custom statuses_

---

_Additional context for reviewer_
None

---
_Release Notes_: 
zendesk:
* support deploy of custom statuses

---
_User Notifications_: 
zendesk:
* the instance default_custom_statuses will be added. It contains the default status for each of the status categories.

### CODE REVIEW

File: packages/zendesk-adapter/test/change_validators/custom_status_active_default.test.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  CUSTOM_STATUS_TYPE_NAME,
+  DEFAULT_CUSTOM_STATUSES_TYPE_NAME, HOLD_CATEGORY, OPEN_CATEGORY,
+  PENDING_CATEGORY, SOLVED_CATEGORY,
+  ZENDESK,
+} from '../../src/constants'
+import { customStatusActiveDefaultValidator } from '../../src/change_validators'
+
+describe('customStatusActiveDefaultValidator', () => {
+  const customStatusType = new ObjectType({ elemID: new ElemID(ZENDESK, CUSTOM_STATUS_TYPE_NAME) })
+  const createStatus = (category: string, isActive: boolean, id: number): InstanceElement => new InstanceElement(
```
Comment (shir-reifenberg) [inline packages/zendesk-adapter/test/change_validators/custom_status_active_default.test.ts:28]: Can we reuse `createStatus` func (at least across the validators)?

File: packages/zendesk-adapter/src/filters/custom_statuses.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+const { RECORDS_PATH } = elementsUtils
+const log = logger(module)
+
+/**
+ * since end_user_label and agent_label must be in deploy and only the raw fields appear in the nacl, the values of the
+ * non-raw fields need to be updated to have the same value of the raw fields.
+ */
+const alignNonRawWithRaw = (change: Change<InstanceElement>): void => {
+  getChangeData(change).value.end_user_label = getChangeData(change).value.raw_end_user_label
```
Comment (shir-reifenberg) [inline packages/zendesk-adapter/src/filters/custom_statuses.ts:46]: nit - 

```suggestion
const { value } = getChangeData(change)
value.end_user_label = value.raw_end_user_label
...
```

File: packages/zendesk-adapter/src/filters/custom_statuses.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+      .filter(inst => inst.value.default)
+
+    const pending = defaultCustomStatuses.find(inst => inst.value.status_category === PENDING_CATEGORY)
+    const solved = defaultCustomStatuses.find(inst => inst.value.status_category === SOLVED_CATEGORY)
+    const open = defaultCustomStatuses.find(inst => inst.value.status_category === OPEN_CATEGORY)
+    const hold = defaultCustomStatuses.find(inst => inst.value.status_category === HOLD_CATEGORY)
+
+    if (pending === undefined || solved === undefined || open === undefined || hold === undefined) {
+      log.error('could not find default status for one of the status categories ')
+      return
```
Comment (shir-reifenberg) [inline packages/zendesk-adapter/src/filters/custom_statuses.ts:66]: It seems like `hold` ticket status is optional (https://support.zendesk.com/hc/en-us/articles/4408889282458) and must be enabled in the account in order to get it. So I think we should still create `defaultCustomStatusesType` if it's missing.
We might have information about if it's enabled or not in `account_features` settings instance so it's might worth checking if we can use this.

**Review by shir-reifenberg — COMMENTED** [review]
Great work! added a few comments and questions
The only major thing is that it seems like `hold` is an optional category (added elaborated comment on this in the relevant file).
Also, as the custom status feature[ needs to be enabled](https://support.zendesk.com/hc/en-us/articles/4412575841306) to use custom statuses, I think we should add a change validator that confirms it's enabled in the target account before deployment, but this can be done in a different pr :)

File: packages/zendesk-adapter/src/filters/field_references.ts [hunk 4]
```diff
@@ -782,6 +787,31 @@ const firstIterationFieldNameToTypeMappingDefs: ZendeskFieldReferenceDefinition[
     serializationStrategy: 'id',
     target: { type: 'section' },
   },
+  {
+    src: { field: 'parent_section_id', parentTypes: ['section'] },
+    serializationStrategy: 'id',
+    target: { type: 'section' },
+  },
```
[the remark this answers is not in the document]
Comment (edenhassid) [reply packages/zendesk-adapter/src/filters/field_references.ts:794]: no it was added by mistake, I will delete it :)

File: packages/zendesk-adapter/test/change_validators/custom_status_active_default.test.ts (hunk 1 above)
Comment (edenhassid) [reply packages/zendesk-adapter/test/change_validators/custom_status_active_default.test.ts:28]: discussed IRL

File: packages/zendesk-adapter/src/filters/custom_statuses.ts (hunk 3 above)
Comment (edenhassid) [reply packages/zendesk-adapter/src/filters/custom_statuses.ts:66]: when hold is not enabled the default of hold can be inactive

Comment (edenhassid) [reply packages/zendesk-adapter/src/filters/custom_statuses.ts:66]: even when custom statuses are disabled we are still able to fetch them and the we have defaults and active statuses.

**Review by shir-reifenberg — APPROVED** [review]
LGTM :)
Still one more question left from previous cycle but pre approving
please don't forget to update the user notifications 🙏

**Pull request merged**